lib/krb5/test_get_creds.c

```diff
@@ -0,0 +1,197 @@
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "krb5_locl.h"
+#include <err.h>
+#include <getarg.h>
+#include <string.h>
+#include <time.h>
+
+static int usage_flag;
+static char *service_name, *ccache_name;
+static struct getargs args[] = {
+    { "service",	's',	arg_string,	&service_name,	NULL,	NULL },
+    { "ccache",		'c',	arg_string,	&ccache_name,	NULL,	NULL },
+    { "version",	0,	arg_flag,	&usage_flag,	NULL,	NULL },
+    { "help",		0,	arg_flag,	&usage_flag,	NULL,	NULL }
+};
+static int const num_args = sizeof(args) / sizeof(args[0]);
+
+static krb5_error_code
+get_service_ticket(
+    krb5_context context,
+    krb5_ccache cc,
+    krb5_principal service,
+    krb5_deltat req_lifetime,
+    int cache_only,
+    krb5_creds **out)
+{
+    krb5_error_code ret;
+    krb5_get_creds_opt opt;
+
+    ret = krb5_get_creds_opt_alloc(context, &opt);
+    if (ret)
+	krb5_err(context, 1, ret, "krb5_get_creds_opt_alloc");
+
+    if (req_lifetime > 0)
+	krb5_get_creds_opt_set_lifetime(context, opt, req_lifetime);
+
+    if (cache_only)
+	krb5_get_creds_opt_set_options(context, opt, KRB5_GC_CACHED);
+
+    ret = krb5_get_creds(context, opt, cc, service, out);
+
+    krb5_get_creds_opt_free(context, opt);
+
+    return ret;
+}
+
+static void
+test_tgs_lifetime(
+    krb5_context context,
+    krb5_ccache cc,
+    krb5_principal service,
+    krb5_deltat req_lifetime,
+    int expect_cached)
+{
+    krb5_error_code ret;
+    krb5_creds *out = NULL;
+    krb5_timestamp now;
+    char expiry_str[26];
+    char expected_str[26];
+    char *tmp;
+    time_t earliest_expected_expiry;
+    krb5_deltat const slop = 60; /* for tolerance */
+
+    /* First, check ccache. */
+    ret = get_service_ticket(context, cc, service, req_lifetime, 1, &out);
+
+    if (expect_cached && ret == KRB5_CC_NOTFOUND)
+	krb5_errx(context, 1, "Expected to find ticket in ccache.");
+    else if (!expect_cached && ret == KRB5_CC_NOTFOUND)
+	; /* This is as expected. */
+    else if (!expect_cached && !ret)
+	krb5_errx(context, 1, "Did not expect to find ticket in ccache.");
+    else if (ret)
+	krb5_err(context, 1, ret, "krb5_get_creds");
+
+    /* If necessary, go to TGS. */
+    if (ret == KRB5_CC_NOTFOUND)
+	ret = get_service_ticket(context, cc, service, req_lifetime, 0,
+	    &out);
+    if (ret)
+	krb5_err(context, 1, ret, "krb5_get_creds");
+
+    ret = krb5_timeofday(context, &now);
+    if (ret)
+	krb5_err(context, 1, ret, "krb5_timeofday");
+
+    earliest_expected_expiry = now + req_lifetime - slop;
+    if (req_lifetime > 0 && out->times.endtime < earliest_expected_expiry)
+    {
+	tmp = ctime(&out->times.endtime);
+	strcpy(expiry_str, tmp);
```

Comment:
Do you want me to change this ctime() thing -- I only put this in to help debug any test failures in future; I could probably instead print out the raw unix timestamps.

Reply:
If you can use `ctime_r()` that would be ideal. Also, indentation should generally match the rest of Heimdal (opening brace on same line).

Reply:
`ctime_r` is not used anywhere else in the codebase (all other places use plain `ctime` instead) and it's not supported by MS Visual C++.  To avoid the debate I've just printed it as raw Unix timestamps.  This only happens as a diagnostic message in a failed test.